components/search/src/filter.rs

```diff
@@ -82,21 +90,26 @@ pub(crate) fn filter_engine_configuration(
     })
 }
 
-fn extract_engine_config(
-    _user_environment: &SearchUserEnvironment,
+fn maybe_extract_engine_config(
+    user_environment: &SearchUserEnvironment,
     record: Box<JSONEngineRecord>,
 ) -> Option<SearchEngineDefinition> {
-    // TODO: Variant handling.
-    Some(SearchEngineDefinition::from_configuration_details(
-        &record.identifier,
-        record.base,
-    ))
+    let base = record.base;
+    record
+        .variants
+        .iter()
+        .rev()
+        .find(|r| matches_user_environment(&r.environment, user_environment))
+        .map(|_variant| {
+            SearchEngineDefinition::from_configuration_details(&record.identifier, base)
+        })
```

Comment:
I left `_variant` in there because we will need to pass `variant` to `from_configuration_details` when I do the next step of handling other items in the variants (partner codes etc). The initial `_` denotes it is unused at the moment. Though I could maybe add a TODO in there.

> I had a thought to not create the base variable, since it's only access once. We could directly do record.base ?

The separate `let base = record.base;` is required because of Rust's borrowing rules / memory management and avoids the clone. I'll forward you a copy of the discussion I had about this. Splitting into two statements doesn't help because we're still hitting the memory issue.

Regarding separating those, I was confused about `.find().map()` until I found out how `map` worked in Rust. So I'm not sure separating those buys us much.

Reply:
> I also learned that if we're not using a parameter, we don't need to name it. So `|_variant|` can become `||`.

Not quite: we can use `|_|` if we don't want to name a parameter, but we can't use `||`. `|_|`, `|_variant|`, and `|variant|` all mean "a closure that takes one argument", but `||` means "a closure that takes zero arguments."

[`Option::map`](https://doc.rust-lang.org/std/option/enum.Option.html#method.map) takes a [closure of type](https://doc.rust-lang.org/reference/types/closure.html) `FnOnce(T) -> U`, but `||` has a type of `FnOnce() -> U`. If we tried to pass `||` to `Option::map`, we'll get a compile error:

```
error[E0593]: closure is expected to take 1 argument, but it takes 0 arguments
  --> src/main.rs:19:10
   |
19 |         .map(|| SearchEngineDefinition::from_configuration_details(&record.identifier, base))
   |          ^^^ -- takes 0 arguments
   |          |
   |          expected closure that takes 1 argument
   |
help: consider changing the closure to take and ignore the expected argument
   |
19 |         .map(|_| SearchEngineDefinition::from_configuration_details(&record.identifier, base))
   |              ~~~
```

> I had a thought to not create the `base` variable, since it's only access once. We could directly do `record.base`?

This is a very subtle one, but directly using `record.base` won't compile:

```
error[E0505]: cannot move out of `record` because it is borrowed
  --> src/main.rs:18:14
   |
10 |       record: Box<JSONEngineRecord>,
   |       ------ binding `record` declared here
11 |   ) -> Option<SearchEngineDefinition> {
12 |       let matching_variant = record
   |  ____________________________-
13 | |         .variants
   | |_________________- borrow of `record.variants` occurs here
...
18 |           .map(|_| SearchEngineDefinition::from_configuration_details(&record.identifier, record.base))
   |            --- ^^^ move out of `record` occurs here                                       ----------- move occurs due to use in closure
   |            |
   |            borrow later used by call
```

@Standard8 asked me about that on Slack, and I actually gave him a partly wrong explanation (it has nothing to do with whether the iterator is consumed or not, but it does have to do with moving `record` into the `.map()` closure while we're borrowing its data. Sorry, Mark! 😬).

I made a [Rust Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=472dc7613212da22ddf4b409fd29472c) with a few stubs, to make sure I also understood the behavior. Here's what's happening:

* `record.variants.iter()` creates an iterator of type `Iterator<Item = &JSONEngineVariant>`, meaning the iterator borrows `record.variants`, and yields `&JSONEngineVariant`s. The `&JSONEngineVariant`s are borrowed references to data from `record.variants`, so `record.variants` can't be moved or dropped until all `&JSONEngineVariant` references are dropped.

* `SearchEngineDefinition::from_configuration_details()` takes an owned `JSONRecordBase`, meaning `base` needs to be moved out of `record`. For that to work, the compiler implicitly moves `record` into the closure that we pass to `.map()`. But...

* 💥 `matching_variant` is an `Option<&JSONEngineVariant>`—it's borrowing data from `record.variants`—so `record.variants` must be valid until _after_ `.map()` returns. But the compiler implicitly moves `record` into the closure, which happens _before_ `.map()` is called. This would invalidate the option's `&JSONEngineVariant` reference, so the borrow checker rejects the code.

We can work around that in a few ways:

1. Move `base` out of `record`, before borrowing `record.variants`. That's what the `base` variable does.
2. Clone the `Option<&JSONEngineVariant>`, so that it becomes an `Option<JSONEngineVariant>`, and avoids the borrow before move: `.rev().find(...).cloned()`.
3. Clone `record.variants`, and turn it into an iterator of owned variants (`Iterator<Item = JSONEngineVariant>`): `record.variants.clone().into_iter().rev()...`.
4. Clone `record.base`, which @Standard8 had originally: `SearchEngineDefinition::from_configuration_details(&record.identifier, record.base.clone())`.

All of these are OK, but (1) avoids an extra allocation, even though it looks a little weird. Another version of (1) is to explicitly move all the fields we need out of `record`, instead of just `base`:

```rust
let JSONEngineRecord {
    identifier,
    variants,
    base,
} = *record;
let matching_variant = variants
    .into_iter()
    .rev()
    .find(|variant| matches_user_environment(&variant.environment, user_environment));
matching_variant.map(|_| SearchEngineDefinition::from_configuration_details(&identifier, base))
```

I included the other options in the Playground link also! 😊